### ISSUE TITLE
single_match should be smarter for known enums

### ISSUE THREAD

**Issue by Manishearth** [Manishearth/rust-clippy#327]
```
    match my_option {
      Some(an_int) => println!("{}", an_int),
      None => {}
    }
```

this can be replaced with an if let, but we don't suggest it, because there's no wildcard. This makes sense for arbitrary enums -- replacing an exhaustive match with an if let can be problematic if the original enum gets updated. But for known enums like Option/Result/Cow, this isn't an issue.

**Comment by llogiq** [comment]
With `Cow`, the cases using only one match arm will be rare anyway. Regarding `Option`/`Result`, I agree. Perhaps I can add the `OptionBool` type from my own [optional](https://github.com/llogiq/optional) crate to the list? I'm positive that it won't get additional variants in the future :smile:

**Comment by Manishearth** [comment]
Yeah, basically anything where we're sure that it won't be extended.

**Comment by llogiq** [comment]
The implementation will need to call `expr_ty` on the matched expression, then use `utils::match_ty` to get the type, then loop to match the type against a white-list of known two-variant enums.

**Comment by mcarton** [comment]
Closing as fixed in #547.
